### ISSUE TITLE
portal can't check for inherited interfaces

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2523]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2523|
|Type|defect|
|Created|2007-03-14 13:52:20Z|

It's something that showed up during the resolution of [#2015](https://github.com/twisted/twisted/issues/2015).

What's the problem? Think of this: 
```
class MyIUserNameInterface(IUsernameHashedPassword)
    pass

class Username(object):
    implements(MyIUserNameInterface)
```

The way portal checks for interfaces breaks the inheritance because it checks:
```
IUsernameHashedPassword in providedBy(Username())
```
which is false, instead of
```
IUsernameHashedPassword.providedBy(Username())
```

which is True.

Patch attached corrects the problem.

Attachments:

* [portal_2523.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/225/22531f5cfa99137a649fc1cf8ec78076341d75ae/1e7a8fb3667d8a2df39e37ab46370b242c2b9a04.diff) (1016 bytes) - added by therve on 2007-03-14 13:52:58Z - 
* [portal-2523-new.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/225/22531f5cfa99137a649fc1cf8ec78076341d75ae/9f859b5950ac481fcd622f41c348cb66b02f25d1.diff) (3957 bytes) - added by antoine on 2007-03-14 16:50:41Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2523 2523
type__defect defect
reporter__therve therve
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1173880340000000 1173880340000000
changetime__1174157993000000 1174157993000000
version__None None
owner__ 
cc__antoine cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_1"></a>|@pitrou commented|
|-|-|

For example, this patch could allow inheriting IUsernamePassword from IUsernameHashedPassword, which seems right semantically (the former provides a superset of the latter).

**Comment by twisted-trac** [comment]
|[<img alt="pitrou's avatar" src="https://avatars.githubusercontent.com/u/1721820?s=50" width="50" height="50">](https://github.com/pitrou)<a name="note_2"></a>|@pitrou commented|
|-|-|

Added new patch: same as Thomas' patch but with test case added.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

Ready to review in portal-inherited-interfaces-2523.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @therve|
|-|-|

This looks like a straightforward correction of a misunderstanding of Zope Interface from the time this code was written.  Go ahead and commit.

As long as you're fixing misunderstandings of ZI though, the docstring you re-indented in ```portal.py``` should say "which provides", not "implementing".  Feel free to fix that before committing.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set status to `closed`|
|-|-|

(In [19847]) Merge portal-inherited-interfaces-2523

Authors: antoine, therve
Reviewer: glyph
Fixes [#2523](https://github.com/twisted/twisted/issues/2523)

portal previously relied on zope.interface.providedBy to give the interfaces
provided by a class, instead of using Interface.providedBy: this particulary
breaks interface inheritance. The branch fixes this and add tests for this
situation.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
